command IDL property reflects invalid values

lukewarlow opened the pull request:
#### 72347c4a2ac0df01f6d4bdf0c3d1d2e3e0215d1b
<pre>
command IDL property reflects invalid values
<a href="https://bugs.webkit.org/show_bug.cgi?id=287189">https://bugs.webkit.org/show_bug.cgi?id=287189</a>

Reviewed by Darin Adler.

Swaps the command IDL property to use manual reflection rather than
the Reflect IDL extended attribute.

Invalid values are now correctly reflected as an empty string.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/the-button-element/command-and-commandfor/command-reflection.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/the-button-element/command-and-commandfor/interface.tentative-expected.txt:
* Source/WebCore/html/HTMLButtonElement.cpp:
(WebCore::togglePopoverAtom):
(WebCore::showPopoverAtom):
(WebCore::hidePopoverAtom):
(WebCore::closeAtom):
(WebCore::showModalAtom):
(WebCore::HTMLButtonElement::commandType const):
(WebCore::HTMLButtonElement::command const):
(WebCore::HTMLButtonElement::setCommand):
* Source/WebCore/html/HTMLButtonElement.h:
* Source/WebCore/html/HTMLButtonElement.idl:

Canonical link: <a href="https://commits.webkit.org/291954@main">https://commits.webkit.org/291954@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/433fa5777b3af2ac3e8c11e9019f730100042984

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94509 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/14101 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3894 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99528 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/45025 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/96559 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14399 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22529 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72116 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29433 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97511 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10717 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/85331 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52448 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10410 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/3029 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/44348 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80630 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/3135 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101572 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21565 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15732 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/81119 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21815 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81351 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80495 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20067 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25047 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/2428 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/14788 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21542 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26677 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/21221 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24684 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22959 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/html/HTMLButtonElement.h:
```diff
@@ -40,6 +40,9 @@ class HTMLButtonElement final : public HTMLFormControlElement {
     const AtomString& value() const;
 
+    const AtomString& command() const;
+    void setCommand(const AtomString& value);
```
Seems unfortunate that we have to write the `setCommand` function even though it’s what `[Reflect]` would have generated. Seems a common pattern to need getter logic, but have the setter side be standard reflection, and I’m surprised we never made a version of `Reflect` for that. Since we do have to write the function, I'll note that the setter doesn’t need to name its argument here:
```suggestion
    void setCommand(const AtomString&);
```

lukewarlow replied:
Yeah a ReflectSetter idl attribute would be very nice and I imagine probably not too much work on the binding side. I know one of the other engines has it (can't think which off the top of my head)

darinadler replied:
I looked over the bindings implementation and indeed it would be an easy project. I might do it soon.

lukewarlow committed:
command IDL property reflects invalid values
https://bugs.webkit.org/show_bug.cgi?id=287189

Reviewed by Darin Adler.

Swaps the command IDL property to use manual reflection rather than
the Reflect IDL extended attribute.

Invalid values are now correctly reflected as an empty string.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/the-button-element/command-and-commandfor/command-reflection.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/the-button-element/command-and-commandfor/interface.tentative-expected.txt:
* Source/WebCore/html/HTMLButtonElement.cpp:
(WebCore::togglePopoverAtom):
(WebCore::showPopoverAtom):
(WebCore::hidePopoverAtom):
(WebCore::closeAtom):
(WebCore::showModalAtom):
(WebCore::HTMLButtonElement::commandType const):
(WebCore::HTMLButtonElement::command const):
(WebCore::HTMLButtonElement::setCommand):
* Source/WebCore/html/HTMLButtonElement.h:
* Source/WebCore/html/HTMLButtonElement.idl:

Canonical link: https://commits.webkit.org/291954@main
